### PR TITLE
SNOW-824882 SSO token cache

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-jy**
### Description
Please explain the changes you made here.
- [x] Added id_token authentication
- [x] Added secureStorage feature
- [x] Added reauthentication  

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [inline lib/authentication/secure_storage/json_credential_manager.js:33]: Our goal here was to verify that only user has permission to this file. What about using fs.constants.S_IRUSR and fs.constants.S_IWUSR

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [inline lib/authentication/secure_storage/json_credential_manager.js:74]: if file exists after 'writeFileSync' there will be updated data and access 600 even if didn't before?

File: lib/core.js
Comment (sfc-gh-pmotacki) [inline lib/core.js:214]: Do we create a log that customer will be use customCredentialManager?

Comment (sfc-gh-pmotacki) [reply lib/core.js:214]: Ok. I see, we have below

File: test/integration/testManualConnection.js
Comment (sfc-gh-pmotacki) [inline test/integration/testManualConnection.js:87]: Could we verify that file contains expected key and some token?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:16]: should it be an environment variable or can we make it parameter configured via `snowflake.configure(...)`?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:27]: should we return null or throw an error?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:31]: should we use sync methods? can we switch to async ones?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:36]: should we only log on info level and return null? should we throw the exception?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:47]: should we log before returning null? should we throw the exception? my question about null vs exception is also for following code

File: lib/connection/connection_config.js
Comment (sfc-gh-dprzybysz) [inline lib/connection/connection_config.js:478]: is the default correct? don't we want to use the temporal storage always? @sfc-gh-pmotacki

File: lib/global_config.js
Comment (sfc-gh-dprzybysz) [inline lib/global_config.js:276]: for MFA tokens we will use the same class? should we pass the directory here?

File: lib/services/sf.js
Comment (sfc-gh-dprzybysz) [inline lib/services/sf.js:633]: are the called method async? I think I saw it without async and with `fs.xyzSync` methods used?

File: test/integration/testManualConnection.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:86]: should we call the functions from `GlobalConfig.getCredentialManager()` with await?

File: test/integration/testManualConnection.js
Comment (sfc-gh-dprzybysz) [inline test/integration/testManualConnection.js:94]: should we await here?

File: test/unit/authentication/custom_credential_manager_test.js
Comment (sfc-gh-dprzybysz) [inline test/unit/authentication/custom_credential_manager_test.js:20]: don't we expect these functions to return promise? (or being async?)

Comment (sfc-gh-ext-simba-jy) [reply test/unit/authentication/custom_credential_manager_test.js:20]: There are two test case scenarios. 1. Sync 2. Async.
You can find the async testing case at the bottom: https://github.com/snowflakedb/snowflake-connector-nodejs/pull/820/files#diff-c92bed55d478910c2b3d0fa5930295b793b9d866ee42e7411ad6ac4fc15531adR66

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:16]: I think that will be better for customers. The reason why I used it is the JDBC is using process.env.SF_TEMPORARY_CREDENTIAL_CACHE_DIR parameter. I will fix it.

File: lib/global_config.js
Comment (sfc-gh-ext-simba-jy) [reply lib/global_config.js:276]: Yes, the original idea is to use the same file. I am not sure whether we need to create a different credential file or not.
Do you think that separating MFA tokens and ID Tokens is better? I agree with this.

File: lib/services/sf.js
Comment (sfc-gh-ext-simba-jy) [reply lib/services/sf.js:633]: Yes, I created the JSON credential manager with sync methods. However, we need to consider a scenario when users use the async functions for the custom credential manager. That is why I added await or then keywords.

File: lib/connection/connection_config.js
Comment (sfc-gh-ext-simba-jy) [reply lib/connection/connection_config.js:478]: On the risk assessment, they mentioned that this option is disabled by default. The customer should read the descripiton first, and need to know what the risk is.

File: test/integration/testManualConnection.js
Comment (sfc-gh-ext-simba-jy) [reply test/integration/testManualConnection.js:87]: This testing is the manual testing. The purpose of this testing is that we need to verify that the ID token connection is working with its own credentials. As a result, we cannot know the expected key and tokens in this testing. I created the unit testing for the JSON credential manager in a different file.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:27]: Users can set the location where they do not have permission. In this case, I am not sure if the driver throws an error. I think disabling the credential cache and going to the next step is a better idea. What do you think?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:31]: If we should use the async functions, I think the sync functions are okay for the condition check. Is there any other reason why we need to use async functions?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:36]: The same idea on above. Please let me know if we need to throw an error in this case or if we just need to disable the credential cache.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:47]: We do not need to log some messages at this point because other error messages are logged already when the driver validates the path on this.getTokenDir. if the tokenDir is null  Also, I am confused about whether we need to throw an exception or disable the credential cache and keep running the driver. Please let me know :)

File: test/integration/testManualConnection.js
Comment (sfc-gh-ext-simba-jy) [reply test/integration/testManualConnection.js:86]: I fixed it

File: test/integration/testManualConnection.js
Comment (sfc-gh-ext-simba-jy) [reply test/integration/testManualConnection.js:94]: Fixed.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js:33]: In fs.access(Sync), fs.constants.S_IRUSR and fs.constants.S_IWUSR cannot be used because It is out of scope. This function only checks the permission, However, I changed the mode with fs.constants.S_IRUSR and fs.constants.S_IWUSR  when the driver writes or modifies the file.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: Sorry, It is not clear to me. Could you please elaborate on this?

Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js]: It is connected with fs.accessSync(tokenCacheFile, fs.constants.R_OK | fs.constants.W_OK). It should be checked before writing token to file as well.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js:26]: If this is a consequence of an incorrect client configuration, we should report the error.

File: lib/global_config.js
Comment (sfc-gh-pmotacki) [reply lib/global_config.js]: I think it could be the same file. If you see possibility to any conflict we can add type to generated key.

Comment (sfc-gh-ext-simba-jy) [reply lib/global_config.js]: No, there is no conflict because the key is generated based on the combination of the host, username, and token name (ID_TOKEN or MFA_TOKEN). The same file is also okay.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: @sfc-gh-dprzybysz @sfc-gh-pmotacki, I raised a question about this problem at https://github.com/snowflakedb/snowflake-sdks-drivers-issues-teamwork/issues/383. Please have a look.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: fs.constants.R_OK | fs.constants.W_OK means the users should have both the read and write permission. Although we already checked before creating the instance, do we need to validate the permission before writing the cache? Or, do you mean we need to check that the users have the permission of fs.constants.S_IRUSR and fs.constants.S_IWUSR?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js:32]: I will discuss it with security team.

File: lib/core.js
Comment (sfc-gh-pmotacki) [inline lib/core.js:222]: Please remove unused code

Comment (sfc-gh-ext-simba-jy) [reply lib/core.js:222]: Sorry, I was waiting for the confirmation. I wasn't sure whether I could move the cachedir option to the connection config or not because the conversation above was not resolved yet.  I will remove this.

Comment (sfc-gh-ext-simba-jy) [reply lib/core.js]: I removed it.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [reply lib/authentication/secure_storage/json_credential_manager.js:30]: in the nodejs driver we would like to use as much as possible the nonblocking code - let's avoid the sync methods

But if we really really need to have it then let it be cc @sfc-gh-pmotacki

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [reply lib/authentication/secure_storage/json_credential_manager.js]: let's keep one thread for discussing it - I am closing this comment

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js]: What about: 'fs.statSync('./yourfile.txt').mode'?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js]: What about: 'fs.statSync('./yourfile.txt').mode'?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:49]: should we warn here about the missing file?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [reply lib/authentication/secure_storage/json_credential_manager.js]: We are using this function e.g. 
```
const jsonCredential = this.readJsonCredentialFile();
return jsonCredential[key] || null;
```
so here there will be the error when json credentials cannot be read - it's better to throw the exception and handle it in upper layers (or we should return an empty object)

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-dprzybysz) [inline lib/authentication/secure_storage/json_credential_manager.js:89]: there is a missing return in the catch - it's ok to return undefined here?

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-pmotacki) [reply lib/authentication/secure_storage/json_credential_manager.js:30]: I agree. When we create new piece of code the general rule is to write unblocking code. Please replace to async methods where it is possible.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: I added it.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: I replaced all the sync functions to the asnyc functions. Please have a look.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: As I replaced the sync functions to async functions, the error will be handled on the upper layer. Please have a look. Thank you.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: I replaced.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: I refactored this code. Please have a look

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: Done.

File: test/integration/testManualConnection.js
Comment (sfc-gh-pmotacki) [inline test/integration/testManualConnection.js:79]: It shouldn't be extended this way. It affected more tests. Please fix it by 
'const connectionOption = {
        ...connOption.externalBrowser,
        clientStoreTemporaryCredential: true
      }'

Comment (sfc-gh-ext-simba-jy) [reply test/integration/testManualConnection.js]: Fixed.

File: lib/core.js
Comment (sfc-gh-ext-simba-jy) [reply lib/core.js]: Done.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: removed.

File: lib/authentication/secure_storage/json_credential_manager.js
Comment (sfc-gh-ext-simba-jy) [reply lib/authentication/secure_storage/json_credential_manager.js]: Fixed.

**Comment by sfc-gh-pmotacki** [comment]
@sfc-gh-ext-simba-jy please update branch with rebase and update security validation links in prepared in doc as well.

**Review by sfc-gh-pmotacki — CHANGES_REQUESTED** [review]
I found some problem  during running manual test. It must be investigated.

**Review by sfc-gh-pmotacki — APPROVED** [review]
LGTM

**Pull request merged**
